crates/torii/grpc/src/server.rs

```diff
@@ -1,30 +1,169 @@
+use std::collections::{HashSet, VecDeque};
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use std::time::Duration;
+
+use futures::{Future, Stream};
+use futures_util::{FutureExt, StreamExt};
 use sqlx::{Pool, Sqlite};
-use tonic::transport::Server;
-use tonic::{Request, Response, Status};
-use world::world_server::{World, WorldServer};
-use world::{MetaReply, MetaRequest};
+use starknet::core::types::{BlockId, MaybePendingStateUpdate, StateUpdate};
+use starknet::providers::{Provider, ProviderError};
+use starknet_crypto::FieldElement;
+use tokio::sync::mpsc::{Receiver, Sender};
+use tokio::time::sleep;
+use tokio_stream::wrappers::ReceiverStream;
+use tonic::{Request, Response, Status, Streaming};
+use world::world_server::World;
+use world::{MetaRequest, MetaResponse};
+
+use self::world::{SubscribeEntitiesRequest, SubscribeEntitiesResponse};
 
+#[allow(non_snake_case)]
 pub mod world {
     tonic::include_proto!("world");
 }
 
+type GetStateUpdateResult<P> =
+    Result<MaybePendingStateUpdate, ProviderError<<P as Provider>::Error>>;
+type StateUpdateFuture<P> = Pin<Box<dyn Future<Output = GetStateUpdateResult<P>> + Send>>;
+
+pub struct SubscriberManager {
+    /// (list of subscribers, list of storage addresses they care about)
+    subscribers: Vec<(Sender<Result<SubscribeEntitiesResponse, Status>>, HashSet<FieldElement>)>,
+}
+
+/// a service which handles entity subscription requests. it is an endless future where it awaits
+/// for new blocks, fetch its state update, and publish them to the subscribers.
+pub struct EntitySubscriptionService<P: Provider> {
+    /// A channel to communicate with the indexer engine, in order to receive the block number that
+    /// the indexer engine is processing at any moment. This way, we can sync with the indexer and
+    /// request the state update of the current block that the indexer is currently processing.
+    block_rx: Receiver<u64>,
+    /// The Starknet provider.
+    provider: Arc<P>,
+    /// A list of state update futures, each corresponding to a block number that was received from
+    /// the indexer engine.
+    state_update_req_futs: VecDeque<StateUpdateFuture<P>>,
+
+    block_num_queue: Vec<u64>,
+    /// Receive subscribers from gRPC server.
+    /// This receives streams of (sender channel, list of entities to subscribe) tuple
+    subscriber_recv:
+        Receiver<(SubscribeEntitiesRequest, Sender<Result<SubscribeEntitiesResponse, Status>>)>,
+}
+
+impl<P> EntitySubscriptionService<P>
```

Comment:
Lets try to consolidate this functionality with https://github.com/dojoengine/dojo/blob/main/crates/torii/core/src/simple_broker.rs since both the graphql and grpc api will expose the same subscription behavior